packages/transactions/src/lifetime.ts

```diff
@@ -102,3 +106,61 @@ export type SetTransactionLifetimeFromTransactionMessage<
     : TTransaction;
+
+const SYSTEM_PROGRAM_ADDRESS = '11111111111111111111111111111111' as Address;
+
+function compiledInstructionIsAdvanceNonceInstruction(
+    instruction: CompiledTransactionMessage['instructions'][number],
+    staticAddresses: Address[],
+) {
+    return (
+        staticAddresses[instruction.programAddressIndex] === SYSTEM_PROGRAM_ADDRESS &&
+        // Test for `AdvanceNonceAccount` instruction data
+        instruction.data != null &&
+        isAdvanceNonceAccountInstructionData(instruction.data) &&
+        // Test for exactly 3 accounts
+        instruction.accountIndices?.length === 3
+    );
+}
+
+function isAdvanceNonceAccountInstructionData(data: ReadonlyUint8Array): boolean {
+    // AdvanceNonceAccount is the fifth instruction in the System Program (index 4)
+    return data.byteLength === 4 && data[0] === 4 && data[1] === 0 && data[2] === 0 && data[3] === 0;
+}
+
+/**
+ * Check if a compiled instruction is an AdvanceNonceAccount instruction
+ * @param compiledTransactionMessage A compiled transaction message that includes a lifetime token
+ * @returns True if the instruction is an AdvanceNonceAccount instruction, false otherwise
+ * Note that this is less precise than checking a decompiled instruction, as we can't inspect
+ * the address or role of input accounts (which may be in lookup tables). However, this is
+ * sufficient for all valid advance durable nonce instructions.
+ * Note that the program address must not be in a lookup table, see [this answer on StackExchange](https://solana.stackexchange.com/a/16224/289)
+ * @see {@link isAdvanceNonceAccountInstruction}
+ */
+export function getTransactionLifetimeConstraintFromCompiledTransactionMessage(
+    compiledTransactionMessage: CompiledTransactionMessage & CompiledTransactionMessageWithLifetime,
+): TransactionBlockhashLifetime | TransactionDurableNonceLifetime {
+    const firstInstruction = compiledTransactionMessage.instructions[0];
+    const { staticAccounts } = compiledTransactionMessage;
+
+    // We need to check if the first instruction is an AdvanceNonceAccount instruction
+    if (firstInstruction && compiledInstructionIsAdvanceNonceInstruction(firstInstruction, staticAccounts)) {
+        const nonceAccountAddress = staticAccounts[firstInstruction.accountIndices![0]];
+        if (!nonceAccountAddress) {
+            throw new SolanaError(SOLANA_ERROR__TRANSACTION__NONCE_ACCOUNT_CANNOT_BE_IN_LOOKUP_TABLE, {
+                nonce: compiledTransactionMessage.lifetimeToken,
+            });
+        }
+        return {
+            nonce: compiledTransactionMessage.lifetimeToken as Nonce,
+            nonceAccountAddress,
+        };
+    } else {
+        return {
+            blockhash: compiledTransactionMessage.lifetimeToken as Blockhash,
+            // This is not known from the compiled message, so we set it to the maximum possible value
+            lastValidBlockHeight: 2n ** 64n - 1n,
```

Comment:
Can save everyone a bit of math.
```suggestion
            lastValidBlockHeight: 0xFFFFFFFFFFFFFFFFn,
```